[datetime] fix(DateInput): make popover focusable via keyboard

michael-yx-wu opened the pull request:
#### Related to https://github.com/palantir/blueprint/issues/4812

#### Changes proposed in this pull request:

Pressing Tab when focused on a `DateInput`'s text field will focus the popover content. Previously, pressing tab would focus the next keyboard-focusable element on the page. Since the popover is rendered with a portal by default, this would usually result in the popover closing.

Does _not_ autofocus the popover to continue to allow users to manually type in a date string if they prefer not to use the popover.

#### Screenshot

Before:
![2021-09-22 17 18 08](https://user-images.githubusercontent.com/3681045/134423329-a04c6b2e-a0cc-4cc2-a1a3-c8a42e302dcf.gif)

After:

![2021-09-22 17 16 34](https://user-images.githubusercontent.com/3681045/134423267-4cd91801-6677-43c4-b2ee-8193db6fd796.gif)

Worth noting that the element tab order in the DatePicker seems weird
1. Left/right arrows
2. Month dropdown
3. Year dropdown

I would expect this order instead:
1. Left arrow
2. Month dropdown
3. Year dropdown
4. Right arrow

michael-yx-wu committed:
Make DateInput's popover keyboard-focusable

michael-yx-wu commented on packages/docs-app/src/examples/core-examples/popoverDismissExample.tsx:
```diff
@@ -36,6 +36,8 @@ export class PopoverDismissExample extends React.PureComponent<
         return (
             <Example options={false} {...this.props}>
                 <Popover
+                    autoFocus={false}
+                    enforceFocus={false}
```
This Popover example continuously reopens by design, but now that the `enforceFocus` behavior is fixed, it will reset the scroll position to this part of the page, making it difficult to interact with the other examples. Happy to do in a separate PR if you prefer

adidahiya replied:
yeah I was mostly just punting this problem because it's fixed in the Popover2 example. no big deal, we can make the change here

adidahiya reviewed (commented):
how do you get back to the input after tabbing into the popover? shift+tab doesn't seem to work:

![2021-09-23 12 30 42](https://user-images.githubusercontent.com/723999/134547272-3e7908c7-fa16-4110-b341-b317df619339.gif)

michael-yx-wu committed:
Allow shift+tab to refocus input

michael-yx-wu commented:
Shift+tab interactions should be fixed now:
![2021-09-24 10 55 49](https://user-images.githubusercontent.com/3681045/134696202-4f5d914a-28a8-4769-b270-09ad1cda5ae1.gif)

michael-yx-wu commented on packages/datetime/src/dateInput.tsx:
```diff
@@ -421,6 +425,7 @@ export class DateInput extends AbstractPureComponent2<DateInputProps, IDateInput
             this.handleDateChange(nextDate, true, true);
         } else if (e.which === Keys.TAB && this.state.isOpen) {
             this.getFirstTabbableElement()?.focus();
+            e.preventDefault();
```
Necessary to prevent focusing the 2nd focusable element

adidahiya replied:
sounds like this should be a code comment

```suggestion
            // necessary to prevent focusing the 2nd focusable element
            e.preventDefault();
```

michael-yx-wu replied:
~~I moved some things around and don't need to this anymore~~

just kidding, I still need this. I'll add the comment

michael-yx-wu committed:
Prevent shift+tab from focusing popover

michael-yx-wu committed:
Close popover if shift+tab on input

adidahiya commented on packages/datetime/src/dateInput.tsx:
```diff
@@ -419,15 +423,21 @@ export class DateInput extends AbstractPureComponent2<DateInputProps, IDateInput
-        } else if (e.which === Keys.TAB) {
-            this.setState({ isOpen: false });
+        } else if (e.which === Keys.TAB && e.shiftKey) {
+            this.handleClosePopover();
```
```suggestion
            // close popover on SHIFT+TAB key press
            this.handleClosePopover();
```

adidahiya commented on packages/datetime/src/dateInput.tsx:
```diff
@@ -463,24 +473,33 @@ export class DateInput extends AbstractPureComponent2<DateInputProps, IDateInput
-            this.unregisterPopoverBlurHandler();
+            this.unregisterPopoverBlurHandlers();
+            this.firstTabbableElement = this.getFirstTabbableElement();
+            this.firstTabbableElement?.addEventListener("blur", this.handlePopoverBlur);
```
this doesn't seem exactly right... I feel like the popover should stay open after a SHIFT+TAB keypress on the first tabbable element?

for example, CLICK, TAB, SHIFT+TAB interaction here:

![2021-09-24 12 20 26](https://user-images.githubusercontent.com/723999/134708673-115ded5e-d0ce-425c-8695-0a460412f34c.gif)

adidahiya replied:
I could go either way, though... open to other ideas

adidahiya reviewed (changes requested):
almost there

michael-yx-wu committed:
Focus input when closing DateInput popover

Since the popover is usually open in a portal which makes it one of the
last elements in the DOM, when closing the popover with tab or
shift+tab, focus is likely to go to the browser address bar or an
element at the bottom of the page, respectively.

This change adds invisible divs to the beginning and end of the popover
that, when focused, return focus to the input and close the popover. We
cannot use the existing Overlay#shouldReturnFocusOnClose prop because it
would reopen the popover.

michael-yx-wu commented on packages/datetime/src/dateInput.tsx:
```diff
@@ -419,70 +424,55 @@ export class DateInput extends AbstractPureComponent2<DateInputProps, IDateInput
+    private getKeyboardFocusableElements = (): HTMLElement[] => {
+        const elements: HTMLElement[] = Array.from(
+            this.popoverContentElement?.querySelectorAll(
+                "button:not([disabled]),input,[tabindex]:not([tabindex='-1'])",
```
This is a subset of the selector used in `Overlay`. I wonder if it's worth sharing that in the future

michael-yx-wu commented:
Makes sense. I'm now keeping the popover open when switching focus to the input. Also fixed the issue of tabbing in the popover focusing the browser address bar.

![2021-09-24 12 55 42](https://user-images.githubusercontent.com/3681045/134712957-53e9c72b-00b6-40a0-880b-1d050b5c3c42.gif)

michael-yx-wu commented:
Will fix tests

michael-yx-wu committed:
Skip failing test

michael-yx-wu commented on packages/datetime/test/dateInputTests.tsx:
```diff
@@ -92,7 +92,11 @@ describe("<DateInput>", () => {
     });
 
-    it("Popover closes when tabbing on first day of the month", () => {
+    // Skipping because simulate just invokes the function passed to React's "on<EventName>" prop
+    // and doesn't actually simulate anything. Properly testing would require running with an actual
+    // browser and focusing specific elements via the DOM API. This would require changing the Karma
+    // config to run with Chrome instead of ChromeHeadless.
+    it.skip("Popover closes when tabbing on first day of the month", () => {
```
I do have a working local version but could only get it to work by switching the Karma browser from `ChromeHeadless` to `Chrome`. The problem with simulate is that there's no way to get the `focusin` event handler functions to run because React doesn't provide an `onFocusIn` prop, only `onFocus`. If I use the DOM API instead and call `Element#focus()`, the browser will fire a `focusin` event on my behalf. I'm not sure if that's an anti-pattern for the unit tests.

adidahiya replied:
ok, this is fine to skip for now. testing focus behavior is tricky in these test suites

adidahiya commented on packages/datetime/src/dateInput.tsx:
```diff
@@ -301,6 +305,8 @@ export class DateInput extends AbstractPureComponent2<DateInputProps, IDateInput
         const { popoverProps = {} } = this.props;
         popoverProps.onClose?.(e);
         this.setState({ isOpen: false });
+        this.startFocusBoundaryElement?.removeEventListener("focusin", this.handleStartFocusBoundaryFocusIn);
```
why do we need to use this ref approach and manually bind focusin event handlers?  why can’t it be declarative like normal React code `<div onFocusIn={this.handleEndFocusBoundaryFocusIn} />` ?

michael-yx-wu replied:
React doesn't support `onFocusIn`, but it looks like their `onFocus` prop listens to the `focusin` event under the hood (https://github.com/facebook/react/issues/6410#issuecomment-671499208). I switched to this approach and everything still works as expected.

michael-yx-wu committed:
Use focus prop and stop manually adding event listener
